s2spy/preprocess.py

```diff
@@ -0,0 +1,220 @@
+        )
+
+
+class Preprocessor:
+    """Preprocessor for s2s data."""
+    def __init__(
+        self,
+        rolling_window_size: int,
+        rolling_min_periods: int = 1,
+        detrend: Union[str, None] = "linear",
+        remove_climatology: bool = True,
+    ):
+        """Preprocessor for s2s data. Can detrend as well as deseasonalize.
+
+        On calling `.fit(data)`, the preprocessor will:
+         - Calculate the rolling mean of the input data.
+         - Calculate and store the trend of the rolling mean.
+         - Calculate and store the climatology of the rolling mean.
+
+        When calling `.transform(data)`, the preprocessor will:
+         - Remove the (stored) trend from a copy of the data.
+         - Remove the climatology from this detrended data.
+         - Return the detrended and deseasonalized data.
+
+        Args:
+            rolling_window_size: The size of the rolling window that will be applied
+                before calculating the trend and climatology. Setting this to 1 will
+                effectively skip this step.
+            rolling_min_periods: The minimum number of periods within a rolling window.
+                If higher than 1 (the default), NaN values will be present at the start
+                and end of the preprocessed data.
+            detrend: Which method to use for detrending. Currently the only method
+                supported is "linear". If you want to skip detrending, set this to None.
+            remove_climatology (optional): If you want to calculate and remove the
+                climatology of the data. Defaults to True.
+        """
+        self._window_size = rolling_window_size
+        self._min_periods = rolling_min_periods
+        self._detrend = detrend
+        self._remove_climatology = remove_climatology
+
+        self._climatology: Union[xr.DataArray, xr.Dataset]
+        self._trend: dict
+        self._is_fit = False
+
+    def fit(self, data: Union[xr.DataArray, xr.Dataset]) -> None:
+        """Fit this Preprocessor to input data.
+
+        Args:
+            data: Input data to preprocess.
+
+        Returns:
+            data_rolling: Same as input but with rolling mean applied.
+        """
+        _check_input_data(data)
+
+        data_rolling = data.rolling(
+            dim={"time": self._window_size}, min_periods=self._min_periods, center=True
+        ).mean()
+        # TODO: give option to be a gaussian-like window, instead of a block.
+
+        if self._detrend is not None:
+            self._trend = _get_trend(data_rolling, self._detrend)
+
+        if self._remove_climatology:
+            self._climatology = _get_climatology(
+                (
+                    _apply_trend(data_rolling, self._detrend, self._trend)
+                    if self._detrend is not None
+                    else data_rolling
+                )
+            )
+        self._is_fit = True
+
+        return data_rolling
```

Comment:
@geek-yang `.fit` should not return data. Additionally, the rolling mean should only be used to calculate the climatology (and can be used for the trend, but that doesn't really matter). 
A rolling mean should not be applied to data with `.transform`.

See the docstring I wrote for the __init__ method.

Reply:
We should also not store the rolling mean (just use it for the climatology, then discard it), to avoid unnecessary memory use.

Reply:
Thanks for your explanation @BSchilperoort . But if so the user cannot get the rolling mean data during preprocessing. There will be cases that the user want rolling mean rather than the raw data (e.g. in my work related to energy transport, I need to take the rolling mean first and use it since then. The rest of the work has nothing to do with the raw input since it is too noisy), therefore I think it is necessary to provide this option. We can talk about it in next meeting.

Reply:
To conclude from what we've discussed today, we should not return anything from `fit` as it breaks the pipeline. Rolling mean should either be handled by another function specific for rolling mean with `fit` and `transform`, or by the users themselves (as it is just a simple operation within the xarray). Sorry for the confusion. The changes are reverted now.